### PR TITLE
Improve documentation for Marshaler

### PR DESCRIPTION
**Pull request by akshayjshah**
Clarify the GoDoc comments and add an example for Marshaler.

### CODE REVIEW

File: marshaler_example_test.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	Name string `json:"name"`
+	Age  int    `json:"age"`
+	Auth Auth   `auth:"auth"`
+}
+
+func (u User) MarshalLog(kv zap.KeyValue) error {
+	kv.AddString("name", u.Name)
+	kv.AddInt("age", u.Age)
+
+	// Put the authentication information in a nested object.
```
Comment (prashantv) [inline marshaler_example_test.go]: After seeing this example, I feel like `KeyValue` maybe should just have `AddObject`. We should chat about the benefits of `Nest`

File: marshaler_example_test.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	defer kv.Nest("auth").CloseField()
+	if err := u.Auth.MarshalLog(kv); err != nil {
```
Comment (prashantv) [inline marshaler_example_test.go]: optionally: `return u.Auth.MarshalLog(kv)`

**Comment by prashantv** [comment]
lgtm, the discussion of `Nest` vs `AddObject` can be tackled separately.

File: marshaler_example_test.go (hunk 1 above)
Comment (akshayjshah) [inline marshaler_example_test.go]: We can definitely add an `AddObject` method, but I think we'll still need `Nest` - not all embedded structs are defined by the user, so I think we still need some way to create a nested namespace for non-`Marshaler`s.

File: marshaler_example_test.go (hunk 2 above)
Comment (akshayjshah) [inline marshaler_example_test.go]: good point.

**Pull request merged**